robotPositioner/robotPositioner.h

```diff
@@ -0,0 +1,382 @@
+        const cv::Size unwrapSize(180,50); // calculate camera input dimensions
+
+        const cv::Mat bubblescopeMask = See3CAM_CU40::createBubblescopeMask(m_cam->getSuperPixelSize());
+        m_cam->autoExposure(bubblescopeMask);
+        m_unwrapper = m_cam->createUnwrapper(unwrapSize);
+    }
+
+    // gets vicon data
+    double getViconData() {
+
+        auto objectData = m_vicon->getObjectData(0);
+   
+        BoBRobotics::Vector3<millimeter_t> translation = objectData.getPosition();
+        BoBRobotics::Vector3<radian_t> rotation = objectData.getAttitude();
+
+        m_pos_X = translation[0].value();
+        m_pos_Y = translation[1].value();
+        m_heading = rotation[0].value() * 180.0/PI;
+    }
+
+
+    // updates the range and bearing from the goal location
+    void updateRangeAndBearing() {
+
+        getViconData(); // update variables
+
+        double delta_x = m_goalPositionX - m_pos_X;
+        double delta_y = m_goalPositionY - m_pos_Y;
+        std::vector<double> delta;
+        delta.push_back(delta_x);
+        delta.push_back(delta_y);
+
+        // calculating the distance form the goal
+        double q = std::inner_product(std::begin(delta), std::end(delta), std::begin(delta), 0.0);
+        m_distanceFromGoal = sqrt(q);
+        m_bearingFromGoal = atan2(delta_y,delta_x)*180.0/PI - m_heading - 90.0;
+        
+        // changing from <0,360> to <-180, 180>
+        m_bearingFromGoal = angleWrapAround(m_bearingFromGoal);
+
+    } 
+
+    template <class T>
+    T angleWrapAround(T angle) {
+        // changing from <0,360> to <-180, 180>
+        if (angle <= 0) { angle += 360; } 
+        if (angle > 180) { angle -= 360; }
+        return angle;
+    }
+
+    template <class T, class S>
+    bool closeTo(const T num, const S threshold) {
+        if (num > -threshold && num < threshold) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+    * reads in the coordinates from a CSV file and puts them in a list
+    */
+    void readCoordinatesFromFile(const std::string csv_file) {
+        std::ifstream in(csv_file);
+        std::vector<std::vector<double>> fields;
+        if (in) {
+            std::string line;
+            while (getline(in, line)) {
+                std::stringstream sep(line);
+                std::string field;
+                fields.push_back(std::vector<double>());
+                while (getline(sep, field, ',')) {
+                    fields.back().push_back(stof(field));
+                }
+            }
+        }
+
+        // saving coordinates from csv
+        for (auto row : fields) {	    	
+            std::vector<double> temp_vec;
+            for (double field : row) {
+                temp_vec.push_back(field);         	
+            }
+            m_coordinates.push_back(temp_vec);
+        }
+    }
+
+    // drives the robot the desired pose (x,y,angle)
+    void driveRobotToGoalLocation() {
```

Comment:
This API essentially blocks the program until the robot has physically got somewhere which I think is less than ideal! A nicer API might be:
- A function to set the goal position -- 
- A function to call every timestep to update the motors -- essentially the inner loop of this function
- A function to query whether the robot has reached it's goal (this could just be a return value of the update function)

Reply:
:+1: